### PR TITLE
add option for automatic JMX retry

### PR DESCRIPTION
**Pull request by SylvainJuge**
- minor code changes
- add simple retry strategy

## What does this PR do?

Some MBeans throw exceptions when the state they expose isn't completely initialized yet.

When the agent starts, such MBeans will be ignored by the agent as they can't be accessed and the agent requires to get their values on registration to get their type, in particular for composite attributes, this is not something that can't be known before invocation.

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxConfiguration.java [hunk 1]
```diff
@@ -137,4 +141,15 @@ public class JmxConfiguration extends ConfigurationOptionProvider {
     ConfigurationOption<List<JmxMetric>> getCaptureJmxMetrics() {
         return captureJmxMetrics;
     }
+
+    private final ConfigurationOption<TimeDuration> faildRetryInterval = TimeDurationValueConverter.durationOption("m")
+        .key("jmx_failed_retry_interval")
+        .tags("internal")
+        .description("If set to a value greater or equal to 1m, the agent will retry failed JMX metric registrations.")
+        .addValidator(isNotInRange(TimeDuration.of("1ms"), TimeDuration.of("59s")))
```
Comment (jackshirazi) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxConfiguration.java:149]:
```suggestion
        .addValidator(isNotInRange(TimeDuration.of("1us"), TimeDuration.of("59s")))
```

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java [hunk 2]
```diff
@@ -175,19 +186,48 @@ synchronized void init(final MBeanServer platformMBeanServer) {
...
                 for (JmxMetricRegistration deletedRegistration : removeAll(newRegistrations, oldRegistrations)) {
                     deletedRegistration.unregister(metricRegistry);
                 }
-
             }
         });
-        register(jmxConfiguration.getCaptureJmxMetrics().get(), platformMBeanServer);
+
+        ConfigurationOption<TimeDuration> failedRetryConfig = jmxConfiguration.getFaildRetryInterval();
+        long retryMillis = failedRetryConfig.getValue().getMillis();
```
Comment (jackshirazi) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:209]: scope is inside the if statement, why here?

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java [hunk 3]
```diff
@@ -175,19 +186,48 @@ synchronized void init(final MBeanServer platformMBeanServer) {
...
+        if (!failedRetryConfig.isDefault()) {
+            retryExecutor = ExecutorUtils.createSingleThreadSchedulingDaemonPool("jmx-retry");
```
Comment (jackshirazi) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:211]: shouldn't we first cancel the previous pool if it exists

Comment (SylvainJuge) [reply apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:215]: Good point, fixed in 91f49cf3d

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java (hunk 2 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java]: Fixed in 91f49cf3d

**Comment by v1v** [comment]
run docs-build

**Review by jackshirazi — APPROVED** [review]

**Pull request merged**